### PR TITLE
iOS InContext Email Protection Signup

### PR DESCRIPTION
**Pull request by amddg44**
Task/Issue URL: https://app.asana.com/0/72649045549333/1202940260106121/f
Tech Design URL:
CC:

**Description**:
Adds Email InContext Signup protection to iOS

**Steps to test this PR**:
1. Starting with a clean install, visit `fill.dev/form/registration-email`
2. Confirm the grey dax icon is visible in the email field
3. Tap the text field (not the icon) and confirm the keyboard is displayed
4. Tap the icon and confirm the prompt to sign up for email protection is presented
5. Tap 'x' to dismiss the prompt and confirm the dax icon is still visible & grey
6. Tap the icon again, and this time select "Don't show again"
7. Confirm the dax icon is no longer visible and that refreshing the page does not bring it back
8. In the app go to Settings > Debug and tap the row "Reset Email Protection InContext Signup"
9. Return to `fill.dev/form/registration-email` and refresh the page. The grey dax icon should now be back
10. Tap the icon again, and this time select "Protect my email"
11. Choose "Get Started" to follow the flow to create a new duck address
12. Confirm that at the end of the sign up process, when dismissing the modal either from the 'Done' button or the 'Try it now', the prompt to chose a duck address is presented and that selecting an address populates the web form
13. Sign out and test the sign in flow (including tapping the link from email to launch the app)

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Review by Mcsadmeal — COMMENTED** [review]
I

File: DuckDuckGo/AppDelegate.swift [hunk 1]
```diff
@@ -436,6 +436,14 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
         os_log("App launched with url %s", log: .lifecycleLog, type: .debug, url.absoluteString)
+
+        if url.absoluteString.starts(with: URL.emailProtection.absoluteString),
```
Comment (afterxleep) [inline DuckDuckGo/AppDelegate.swift]: Nit.  WDYT about extracting this logic to a helper function or a url extenstion to keep this method clean?

File: DuckDuckGo/EmailSignupPromptViewController.swift [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        controller.view.backgroundColor = .clear
+        presentationController?.delegate = self
+        installChildViewController(controller)
+    }
+
+}
+
+extension EmailSignupPromptViewController: UISheetPresentationControllerDelegate {
+    func presentationControllerDidDismiss(_ presentationController: UIPresentationController) {
+//            Pixel.fire(pixel: .autofillLoginsPasswordGenerationPromptDismissed)
```
Comment (afterxleep) [inline DuckDuckGo/EmailSignupPromptViewController.swift]: Should this be removed?

File: DuckDuckGo/EmailSignupViewController.swift [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+        return manager
+    }()
+
+    private var url: URL? {
+        didSet {
+            guard let url = url else {
+                navigationItem.rightBarButtonItems = []
+                return
+            }
+            if url.absoluteString.hasSuffix("email/") || url.absoluteString.hasSuffix("email/start-incontext") {
```
Comment (afterxleep) [inline DuckDuckGo/EmailSignupViewController.swift]: These strings can be moved to your constants enum

File: DuckDuckGo/EmailSignupViewController.swift [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+
+    private var canGoBack: Bool {
+        let webViewCanGoBack = webView.canGoBack
+        let navigatedToError = webView.url != nil
+        return webViewCanGoBack || navigatedToError
+    }
+
+    private lazy var backBarButtonItem: UIBarButtonItem = {
+        let button = UIButton(type: .system)
+        button.setImage(UIImage(systemName: "chevron.left"), for: .normal)
```
Comment (afterxleep) [inline DuckDuckGo/EmailSignupViewController.swift]: chevron.left can also be a constant

File: DuckDuckGo/EmailSignupViewController.swift [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+#endif
+    }
+
+    private func webViewUrlHasChanged() {
+        url = webView.url
+    }
+
+    private func setupNavigationBarTitle() {
+        let titleLabel: UILabel = UILabel()
+        titleLabel.text = "DuckDuckGo"
```
Comment (afterxleep) [inline DuckDuckGo/EmailSignupViewController.swift]: Should this be a localized string?  if not, probably a constant.

File: DuckDuckGo/EmailSignupViewController.swift [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(onDuckDuckGoEmailDidCloseEmailProtection),
+                                               name: .emailDidCloseEmailProtection,
+                                               object: nil)
+    }
+
+    @objc
+    private func onDuckDuckGoEmailSignIn(_ notification: Notification) {
+        if signupStage != .complete {
+            // TODO - pixel
```
Comment (afterxleep) [inline DuckDuckGo/EmailSignupViewController.swift]: Assuming this is what's coming up next with the pixels pr, right?

File: DuckDuckGo/AutofillLoginPromptView.swift [hunk 7]
```diff
@@ -93,6 +91,18 @@ struct AutofillLoginPromptView: View {
         }
     }
 
+    private var horizontalPadding: CGFloat {
+        if AutofillViews.isIPhonePortrait(verticalSizeClass, horizontalSizeClass) {
```
Comment (afterxleep) [inline DuckDuckGo/AutofillLoginPromptView.swift]: I think the nested if can be replaced with a guard or even better a switch and a couple cases..

```
        guard AutofillViews.isIPhonePortrait.... else { 
            return Const.Size.closeButtonOffset 
         } 
        
        if isSmallFrame(frame) { 
            return ....
          } 
        
        return ....
```

**Review by afterxleep — APPROVED** [review]
Works as expected.  Left a few nitpicks

**Pull request merged**